### ISSUE TITLE
Chat plugin does not work with zmq-multiplex

### ISSUE THREAD

**Issue by b123400** [SkygearIO/skygear-server#342]
OK, finally get the reason.
Chat plugin uses few after_save hooks with `async=True`, if I set all of them to `async=False` it works fine.
The current protocol assumes requested to be nested, if a child request does not end, the parent request cannot end. This is not true when `async=True`.

I can think of few directions to solve the problem.

1. Force `async=False` when using a zmq-multiplex
2. When `async=True`, execute hooks with another worker (not reusing worker)
    - This means we might go back to the [original issue](https://github.com/SkygearIO/skygear-server/issues/295), workers might be used up when there are a lot of async hooks
3. When `async=True`, execute hooks with newly created worker
    - This requires quite some changes because workers are currently spawned at the plugin side
4. When the plugin is executing a hook with `async=True`, it should internally queue up all messages, reorder them to make them sync again.
    - Better than (1). This way golang will still perform hooks in async, so no slow down at skygear-server side
    - The hooks are sync, arguably a not exactly what `async=True` means, it is a complicated behaviour for user to understand
    - Difficult to maintain

**Comment by b123400** [comment]
I just updated the issue with what I have found in the zmq + chat runtime. Any advise...?
@cheungpat @rickmak @ben181231

**Comment by cheungpat** [comment]
I would opt for choice 4 and I agree with you that it is the most difficult one.

I think we should not allow a request to spawn coroutines that 1) run concurrently, and 2) call cloud code. This will potentially consume workers  in a short amount of time and worker is a limited resource for non-async python.

I think to have hook run `async=True` is for having the hook to run after the database transaction commits and the request does not have to wait for the hook to finish. If I am wrong, others could state what their expectations maybe.

We still need to discuss if you go for option 4 but I am pretty sure we do not break people’s assumption. But option 1 definitely does.

**Comment by rickmak** [comment]
> async=True is for having the hook to run after the database transaction commits and the request does not have to wait for the hook to finish.

Yes. Here we are kind of mix-up concept of blocking and async. Async implies non-blocking. Sync implies blocking.

So for 4, it will purely become non-blocking sync operation. Which may be fine, it may actually don't break how people currently understand and options. And how we describe it to user in [doc](https://docs.skygear.io/guides/cloud-function/database-hooks/python/)

**Comment by b123400** [comment]
> So for 4, it will purely become non-blocking sync operation. Which may be fine, it may actually don't break how people currently understand and options. And how we describe it to user in doc

I think it at least breaks the expectation I have with the name `async`, should be changed to something like `nonblocking=True`?

**Comment by chpapa** [comment]
Or would it be we have some werid assumption with zmq-multiplex in the first place? If multiple requests are sync (thus "parent request" need to wait til "child requests" to finish), I think it is fair to say we want to re-use worker.

However, if the requests are async, it sounds more reasonable to just queue up and have workers handle things in later time. In that case, a workers pool is a more "standard approach"? Just my 2 cents.

**Comment by b123400** [comment]
4 means queuing up tasks. Though, instead of of using the worker like 2, it queues tasks for a single worker.

**Comment by chpapa** [comment]
> 4 means queuing up tasks. Though, instead of of using the worker like 2, it queues tasks for a single worker.

Oh sorry for confusion, my sentence of "it sounds more reasonable to just queue up and have workers handle things in later time" imply it doesn't have to be the same worker.

Reason besides a more "standard approach", is I guess we would actually need to have some sort of dynamic worker pool anyway, so maybe doing that instead of handle the complicated logic of queue up async tasks would make us more flexible in future. Just my 2 cents.

**Comment by rickmak** [comment]
@b123400 May I know how many thread you are using to test the halt. I tried `ZMQ_THREAD_POOL=64` and seem still halt. I suspect it is not about worker exhausted.

> When async=True, execute hooks with another worker (not reusing worker)
> When async=True, execute hooks with newly created worker

If we make zmq transport have dynamic thread as in http implementation at py-skygear, it will solve the deadlock problem.

It may defer the problem to resource shortage as @cheungpat said. Let introduce another kind of warning if the program spawn too many thread?

**Comment by b123400** [comment]
I am using 4 threads, but that doesn't matter. It halts despite how many threads are used. The problem comes from skygear server sending messages to plugin when the plugin is still processing. When the plugin sends a nested request to server, the previous message from server is misunderstood as the response of the request.

**Comment by rickmak** [comment]
> When the plugin sends a nested request to server, the previous message from server is misunderstood as the response of the request.

This sound like a bug that can fix independently regardless of our choice?

Updated: please take a look at https://github.com/SkygearIO/skygear-server/issues/347 Which @Steven-Chan and I spend quit some time to make the minimal reproducible code.

====
For a workable fix, let me propose this:

- When async=True, execute hooks with another worker (not reusing worker)
- Make zmq threading at py-skygear dynamic https://github.com/SkygearIO/py-skygear/issues/134

**Comment by b123400** [comment]
> This sound like a bug that can fix independently regardless of our choice?

I don't think so. To fix this, it will mean implementing 4. Imagine the case:

1. Plugin send `record:save` to server
2. Server trigger after_save hook (`async=False`)
    2.1. Inside the hook, plugin send a request (e.g. `user:query`) to server
    2.2. Server send response for `user:query` (Lets call this ResponseA)
3. Server send response of `record:save`(Lets call this ResponseB)

Because the trigger is async, it is possible that ResponseB comes between 2.1 and 2.2. In this case, I have to queue ResponseB somewhere in the program, then wait for ResponseA. This queue is essentially option 4.

**Comment by rickmak** [comment]
First, in the test provided by Steven https://github.com/SkygearIO/skygear-server/issues/347, there is no 2.1 and 2.2. So you are referring #347 is not the issue you found? If so, can we have a minimal test case like #347 to reproduce the issue?

> Because the trigger is async, it is possible that ResponseB comes between 2.1 and 2.2. In this case, I have to queue ResponseB somewhere in the program, then wait for ResponseA. This queue is essentially option 4.

I don't understand why 3 need to depends on 2. We should treat 2 as an independent request.

**Comment by b123400** [comment]
#347 occurs in the master branch, not zmq-multiplex, therefore is not related to this issue, please don't mix them up.

Minimal code to reproduce:

```
@skygear.op('catapi:get')
def get_cat():
    container = SkygearContainer(api_key='master', user_id='user1')
    result = container.send_action('record:save', {
        'records': [{'_id': 'Note/thing',}],
    })
    return result

@skygear.after_save('Note', async=True)
def Note_async_after_save(record, original_record, conn):
    container = SkygearContainer(api_key='master', user_id='user1')
    result = container.send_action('me', {})
    # There is a chance you get a note here, instead of user
    return result
```

**Comment by b123400** [comment]
Just sent https://github.com/SkygearIO/skygear-server/pull/349 to prevent reusing worker for async hooks.

After fixing the bug there are 2 other bugs that prevented chat from working:
- https://github.com/SkygearIO/skygear-server/issues/347 (Which is fixed via #348)
- Heartbeat is putting worker back into pool even though it is waiting for nested response. (This is fixed in #349 too)
